src/middlewared/middlewared/service/core_service.py

```diff
@@ -172,25 +199,52 @@ def get_tasks(self):
         ),
         register=True,
     ))
-    def get_jobs(self, filters, options):
-        """Get the long running jobs."""
+    @pass_app(rest=True)
+    def get_jobs(self, app, filters, options):
+        """
+        Get information about long-running jobs.
+        If authenticated session does not have the FULL_ADMIN role, only
+        jobs owned by the current authenticated session will be returned.
+        """
+        if self.__is_limited_to_own_jobs(app):
+            username = app.authenticated_credentials.user['username']
+            jobs = list(self.middleware.jobs.for_username(username).values())
+        else:
+            jobs = list(self.middleware.jobs.all().values())
+
         raw_result = options['extra'].get('raw_result', True)
         jobs = filter_list([
-            i.__encode__(raw_result) for i in list(self.middleware.jobs.all().values())
+            i.__encode__(raw_result) for i in jobs
         ], filters, options)
         return jobs
 
+    @no_authz_required
     @accepts(Int('id'))
     @job()
     async def job_wait(self, job, id_):
-        return await job.wrap(self.middleware.jobs[id_])
+        target_job = self.middleware.jobs[id_]
 
+        if job.credentials and job.credentials.is_user_session:
+            this_user = job.credentials.user['username']
+            if not credential_has_full_admin(job.credentials):
+                if (
+                        target_job.credentials is None or
+                        not target_job.credentials.is_user_session or
+                        target_job.credentials.user['username'] != this_user
+                   ):
+                    raise CallError(f'{id_}: job is not owned by current session')
+
+        return await job.wrap(target_job)
+
+    @no_authz_required
     @accepts(Int('id'), Dict(
         'job-update',
         Dict('progress', additional_attrs=True),
     ))
-    def job_update(self, id_, data):
+    @pass_app()
+    def job_update(self, app, id_, data):
         job = self.middleware.jobs[id_]
+        self.__check_job_owned_by_session(job, app)
```

Comment:
This is called by worker from another process pool. It should be private and this check is not needed (no users will update their jobs). `@no_authz_required` should also not be necessary

Reply:
Good to know. I'll back out those changes